overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,381 @@
+			panic("need full first-boot to initialize brand and model of device")
+		}
+	}
+
+	for _, chg := range m.state.Changes() {
+		if chg.Kind() == "become-operational" && !chg.Status().Ready() {
+			// change already in motion
+			return nil
+		}
+	}
+
+	// XXX: some of these will need to be split and use hooks
+	// retries might need to embrace more than one "task" then,
+	// need to be careful
+
+	genKey := m.state.NewTask("generate-device-key", i18n.G("Generate device key"))
+	requestSerial := m.state.NewTask("request-serial", i18n.G("Request device serial"))
+	requestSerial.WaitFor(genKey)
+
+	chg := m.state.NewChange("become-operational", i18n.G("Setting up device identity"))
+	chg.AddAll(state.NewTaskSet(genKey, requestSerial))
+
+	return nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *DeviceManager) Ensure() error {
+	err := m.ensureOperational()
+	if err != nil {
+		return err
+	}
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *DeviceManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *DeviceManager) Stop() {
+	m.runner.Stop()
+}
+
+var (
+	keyLength = 4096
+	// XXX: a 2nd different URL for nonce?
+	// TODO: this will come as config from the gadget snap
+	serialRequestURL = "https://serial.request" // XXX dummy value!
+)
+
+func (m *DeviceManager) doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var keyID string
+	// XXX: make this a field of DeviceState?
+	err := st.Get("device-key-id", &keyID)
+	if err == nil {
+		// nothing to do
+		return nil
+	}
+	if err != state.ErrNoState {
+		return err
+	}
+
+	keyPair, err := rsa.GenerateKey(rand.Reader, keyLength)
+	if err != nil {
+		return fmt.Errorf("cannot generate device key pair: %v", err)
+	}
+
+	privKey := asserts.RSAPrivateKey(keyPair)
+
+	// TODO: simplify key mgmt API? "device" here is a dummy authorityID
+	err = m.keypairMgr.Put("device", privKey)
+	if err != nil {
+		return fmt.Errorf("cannot store device key pair: %v", err)
+	}
+
+	st.Set("device-key-id", privKey.PublicKey().ID())
+	return nil
+}
+
+func (m *DeviceManager) keyPair() (asserts.PrivateKey, error) {
+	var keyID string
+	err := m.state.Get("device-key-id", &keyID)
+	if err == state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot find device key pair")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	privKey, err := m.keypairMgr.Get("device", keyID)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read device key pair: %v", err)
+	}
+	return privKey, nil
+}
+
+type serialSetup struct {
+	SerialRequest string `json:"serial-request"`
+}
+
+type requestIDResp struct {
+	RequestID string `json:"request-id"`
+}
+
+func prepareSerialRequest(t *state.Task, privKey asserts.PrivateKey, device *auth.DeviceState, client *http.Client) (string, error) {
+	st := t.State()
+	st.Unlock()
+	defer st.Lock()
+	resp, err := client.Get(serialRequestURL)
```

Comment:
This assumes the same endpoint is used for getting the request-id and do the serial request, I know we talked that they could be different endpoints. FWIW, in the current server implementation they are different, and I thought we preferred those to be different? Should this be changed server side?